homeassistant/components/simplisafe/config_flow.py

```diff
@@ -19,56 +24,131 @@ class SimpliSafeFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+        """Get an authenticated SimpliSafe API client."""
+        client_id = await async_get_client_id(self.hass)
+        websession = aiohttp_client.async_get_clientsession(self.hass)
+
+        return await API.login_via_credentials(
+            self._username, self._password, client_id=client_id, session=websession,
+        )
+
+    async def _async_login_during_step(self, *, step_id, form_schema):
+        """Attempt to log into the API from within a config flow step."""
+        errors = {}
+
+        try:
+            simplisafe = await self._async_get_simplisafe_api()
+        except PendingAuthorizationError:
+            LOGGER.info("Awaiting confirmation of MFA email click")
+            return await self.async_step_mfa()
+        except InvalidCredentialsError:
+            errors = {"base": "invalid_credentials"}
+        except SimplipyError as err:
+            LOGGER.error("Unknown error while logging into SimpliSafe: %s", err)
+            errors = {"base": "unknown"}
+
+        if errors:
+            return self.async_show_form(
+                step_id=step_id, data_schema=form_schema, errors=errors,
+            )
+
+        return await self.async_step_finish(
+            {
+                CONF_USERNAME: self._username,
+                CONF_TOKEN: simplisafe.refresh_token,
+                CONF_CODE: self._code,
+            }
+        )
+
+    async def async_step_finish(self, user_input=None):
+        """Handle finish config entry setup."""
+        existing_entry = await self.async_set_unique_id(self._username)
+        if existing_entry:
+            self.hass.config_entries.async_update_entry(existing_entry, data=user_input)
+            return self.async_abort(reason="reauth_successful")
+        return self.async_create_entry(title=self._username, data=user_input)
+
     async def async_step_import(self, import_config):
         """Import a config entry from configuration.yaml."""
         return await self.async_step_user(import_config)
 
+    async def async_step_mfa(self, user_input=None):
+        """Handle multi-factor auth confirmation."""
+        if user_input is None:
+            return self.async_show_form(step_id="mfa")
+
+        try:
+            simplisafe = await self._async_get_simplisafe_api()
+        except PendingAuthorizationError:
+            LOGGER.error("Still awaiting confirmation of MFA email click")
+            return self.async_show_form(
+                step_id="mfa", errors={"base": "still_awaiting_mfa"}
+            )
+
+        return await self.async_step_finish(
+            {
+                CONF_USERNAME: self._username,
+                CONF_TOKEN: simplisafe.refresh_token,
+                CONF_CODE: self._code,
+            }
+        )
+
+    async def async_step_reauth(self, user_input=None):
+        """Handle configuration by re-auth."""
+        self._code = user_input.get(CONF_CODE)
+        self._username = user_input[CONF_USERNAME]
```

Comment:
The input here will never be None.
```suggestion
    async def async_step_reauth(self, config):
        """Handle configuration by re-auth."""
        self._code = config.get(CONF_CODE)
        self._username = config[CONF_USERNAME]
```